fasto/Fasto/CodeGen.fs

```diff
@@ -653,8 +653,59 @@ let rec compileExp  (e      : TypedExp)
          counter computed in step (c). You do this of course with a
          `SW(counter_reg, place, 0)` instruction.
   *)
-  | Filter (_, _, _, _) ->
-      failwith "Unimplemented code generation of filter"
+  | Filter (farg, arr_exp, tp, pos) ->
+      let size_reg      = newReg "size_reg"
+      let arr_reg       = newReg "arr_reg"
+      let elem_reg      = newReg "elem_reg"
+      let res_reg       = newReg "res_reg"
+      let addr_reg      = newReg "addr_reg"
+      let i_reg         = newReg "i_reg"
+      let arrh_reg      = newReg "arrh_reg"
+      let counter_reg   = newReg "counter_reg"
+      let tmp_reg       = newReg "tmp_reg"
+
+      let arr_code      = compileExp arr_exp vtable arr_reg
+
+      let get_size      =   [ LW (size_reg, arr_reg, 0);  ]
+      let init_regs     =   [ ADDI (addr_reg, place, 4);
+                              ADDI (arrh_reg, place, 0);
+                              MV (i_reg, Rzero);
+                              MV (counter_reg, Rzero);
+                              ADDI (elem_reg, arr_reg, 4) ]
+
+      let loop_beg      = newLab "loop_beg"
+      let loop_end      = newLab "loop_end"
+      let check_wrong   = newLab "check_wrong"
+      let tmp_reg       = newReg "tmp_reg"
+
+      let loop_header   = [ LABEL (loop_beg);
+                            SUB (tmp_reg, i_reg, size_reg);
+                            BGE (tmp_reg, i_reg, loop_end)  ] 
```

Comment:
You're subtracting n from i here and comparing if it's greater than i. I'm guessing this will cause a endless loop if you run it with rars.